ssc: docker hub image analysis (#16401)
docker/docs: develop/analyze-images/index.md

> ---

Reviewer: toc.yaml need changing? to add this page?

description: Analyze and build secure images, best practices guide
keywords: docker scan, scan, images, vulnerability, analyze, supply chain
redirect_from:
  - /develop/scan-images/
---

This page contains recommendations and best practices for analyzing image
security and building secure images.

Container security spans multiple teams: developers, security, and operations.
That's because there are multiple layers of security that apply to containers:

- The container image and the software running inside
- The interaction between a container, the host operating system, and other
  containers on the same host
- The host operating system
- Container networking and storage

## Analyzing images

With Docker, you can bring supply chain security into the development workflow.
The Docker platform offers an image analysis service that extends the existing,
familiar process of vulnerability detection. Image analysis not only helps you
detect vulnerabilities, but also reveals where each vulnerability comes from,
and provides suggestions on how you can remediate them. The process builds on
continuously indexing the contents of your images, and evaluating that against

> the Docker's image vulnerability database.

Reviewer: ```suggestion
 Docker's image vulnerability database.
```

You can trigger image analysis, view, and inspect vulnerabilities through Docker
Hub and Docker Desktop. For detailed information, see
[Image analysis](../../docker-hub/image-analysis.md).

> > **Note**

Reviewer: Maybe move this right to the top? Consistent with how we have it on other docs pages

>
> Image analysis in Docker Hub is available for developers subscribed to a
> Docker Pro, Team, or a Business tier. For more information about the pricing
> plans, see
> [Docker Pricing](https://www.docker.com/pricing?utm_source=docker&utm_medium=webreferral&utm_campaign=docs_driven_upgrade){:target="_blank"
> rel="noopener" class="_"}.

### View the image report in Docker Desktop

> **Note**
>
> This is an experimental feature.

> Docker Desktop provides you a snapshot of your vulnerabilities status of images.

Reviewer: ```suggestion
Docker Desktop provides a vulnerability status of your images.
```

Pull the image and select it from the **Images** tab in the Docker Dashboard.
The image details view opens, where you can find the image analysis results.

![Analysis report in Docker Desktop](images/desktop.png){:width="700px"}

## Best practices

As a developer, you can take a few simple steps to improve the security of your
container. This includes:

1. Choosing the right base image from a trusted source and keeping it small
2. Using multi-stage builds
3. Rebuilding images
4. Analyzing images during development
5. Analyzing images during production

Now, let’s take a detailed look at each of these best practices:

### Choose the right base image

> The first step towards achieving a secure image is to choose the right base

Reviewer: Is the first sentence needed? I feel it's a repeat of the above

image. When choosing an image, ensure it is built from a trusted source and keep
it small.

Docker Hub has more than 8.3 million repositories. Some of these images are
[Official Images](../../docker-hub/official_images.md), which are published by
Docker as a curated set of Docker open source and drop-in solution repositories.
Docker also offers images that are published by
[Verified Publishers](../../docker-hub/publish/index.md). These high-quality
images are published and maintained by the organizations partnering with Docker,
with Docker verifying the authenticity of the content in their repositories.
When you pick your base image, look out for the **Official Image** and
**Verified Publisher** badges.

![Docker Hub Official and Verified Publisher images](images/hub-official-images.png){:width="700px"}

When building your own image from a Dockerfile, ensure you choose a minimal base

> image that matches your requirements. A smaller base image not just offers

Reviewer: ```suggestion
image that matches your requirements. A smaller base image offers
```

> portability and fast downloads, but also shrinks the size of your image and

Reviewer: ```suggestion
portability and fast downloads, and also shrinks the size of your image and
```

minimizes the number of vulnerabilities introduced through the dependencies.

> We also recommend that you use two types of base images: the first image for

Reviewer: ```suggestion
We also recommend that you use two types of base images. The first image for
```

- Containers should be easy to destroy and rebuild.
- Use a small base image (such as Linux Alpine). Smaller images are easier to
  distribute.
- Avoid installing unnecessary packages. This keeps the image clean and safe.
- Avoid cache hits when building.
- Automatically analyze your image before deploying to avoid pushing vulnerable
  containers to production.
- Analyze your images daily both during development and production for
  vulnerabilities. Automatically rebuild images if necessary.

For detailed best practices and methods for building efficient images, see
[Dockerfile best practices](../develop-images/dockerfile_best-practices.md).

### Analyze images during development

Creating an image from a Dockerfile and even rebuilding an image can introduce
new vulnerabilities in your system. Analyzing your Docker images during
development should be part of your workflow to catch vulnerabilities earlier in
your development. You should analyze images at all stages of the development
cycle, and ideally consider automating analysis. For example, consider
configuring automated analysis during the build process, before pushing the

> image to Docker Hub (or any other registry), and finally before pushing it to a

Reviewer: ```suggestion
image to Docker Hub or any other registry, and finally before pushing it to a
```
